profiling/service/service.go

```diff
@@ -0,0 +1,123 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// The service package defines methods to register a gRPC client/service for a
+// profiling service that is exposed in the same server. This service can be
+// queried by a client to remotely manage the gRPC profiling behaviour of an
+// application.
+package service
+
+import (
+	"context"
+	"errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+
+	"google.golang.org/grpc/internal/profiling"
+	pspb "google.golang.org/grpc/profiling/proto/service"
+)
+
+// Profiling configuration options.
+type ProfilingConfig struct {
+	// Setting this to true will enable profiling.
+	Enabled bool
+
+	// Profiling uses a circular buffer (ring buffer) to store statistics for
+	// only the last few RPCs so that profiling stats do not grow unbounded. This
+	// parameter defines the upper limit on the number of RPCs for which
+	// statistics should be stored at any given time. An average RPC requires
+	// approximately 2-3 KiB of memory for profiling-related statistics, so
+	// choose an appropriate number based on the amount of memory you can afford.
+	StreamStatsSize uint32
+
+	// To expose the profiling service and its methods, a *grpc.Server must be
+	// provided.
+	Server *grpc.Server
+}
+
+var errorNilServer = errors.New("No grpc.Server provided.")
+
+// Init takes a *ProfilingConfig to initialise profiling (turned on/off
```

Comment:
When/Who call this function? I couldn't find a call in any of your PRs. Maybe I missed it. Could you please point me to it.

Reply:
This will be called by the application when they are registering their server, so it's outside gRPC. Later, once all PRs are merged, I'll add a profiling example to better showcase how to register profiling so that gRPC users can read and implement it in their servers. I'll also add documentation, of course.

FWIW, this is what it looks like in my local test example:

```go
	srv := grpc.NewServer()
	one.RegisterOneServer(srv, &oneServer{})
	lis, err := net.Listen("tcp", address)
	if err != nil {
		return err
	}

	profiling.Init(&profiling.ProfilingConfig{Enabled: true, Server: srv})

	go srv.Serve(lis)
```

Reply:
I would prefer exporting the `profilingServer` object and moving stuff from `ProfilingConfig` into it. So, things would become like this:

// Package profiling
```golang
type Server struct {
	Enabled bool
	StreamStatsSize uint32
}
```

// Package callsite (application main)
```golang
...
    srv := grpc.NewServer()
   // Register other server
    pspb.RegisterProfilingServer(srv, &profiling.Server{})
...
```

Let me know if you like this better.

Reply:
Quick question, when would `InitStats` be called?

Reply:
Fair enough :)